File: apps/codemap-search/tests/fixtures/navigation/go/nested.go
package checkout

import "context"

type OrderLine struct {
	Price    int
	Quantity int
}

type OrderDTO struct {
	Total int
}

type Mapper interface {
	MapOrder([]OrderLine) OrderDTO
}

type Repository interface {
	Save(context.Context, OrderDTO) error
}

type Inventory interface {
	Reserve(context.Context, []OrderLine) error
}

type Service struct {
	mapper    Mapper
	repo      Repository
	inventory Inventory
}

func (s *Service) Submit(ctx context.Context, lines []OrderLine) error {
	dto := s.mapper.MapOrder(lines)
	audit := Trace(ctx, "checkout.submit")

	if err := s.inventory.Reserve(ctx, lines); err != nil {
		return err
	}
	if err := s.repo.Save(ctx, dto); err != nil {
		return err
	}
	audit.Done()
	return nil
}

File: apps/codemap-search/tests/fixtures/navigation/rust/nested.rs
use crate::audit::*;
use crate::repo::OrderRepo;

pub struct OrderLine {
    price: i64,
    quantity: i64,
}

pub struct OrderDto {
    total: i64,
}

pub trait Mapper {
    fn map(&self, lines: &[OrderLine]) -> OrderDto;
}

macro_rules! audit_event {
    ($name:expr, $dto:expr) => {
        trace_order($name, $dto.total)
    };
}

pub struct CheckoutService<M, I> {
    mapper: M,
    repo: OrderRepo,
    inventory: I,
}

impl<M, I> CheckoutService<M, I>
where
    M: Mapper,
    I: Inventory,
{
    pub fn submit(&self, lines: &[OrderLine]) -> Result<OrderDto, String> {
        let dto: OrderDto = self.mapper.map(lines);
        let audit = audit_event!("checkout.submit", dto);

        self.inventory.reserve(lines)?;
        self.repo.save(&dto)?;
        audit.finish();
        Ok(dto)
    }
}
